Add EIP: Perpetually Valid Signed Voluntary Exits (#7044)
ethereum/EIPs: EIPS/eip-7044.md

---
eip: 7044
title: Perpetually Valid Signed Voluntary Exits
description: Lock voluntary exit signature domain on capella for perpetual validity
author: Lion (@dapplion)

> discussions-to: https://github.com/ethereum/consensus-specs/pull/3288

Reviewer: you might need to open a discussion or ethereum-magicians.org or https://ethresear.ch

Reviewer 2: You need to create a discussion thread on ethereum-magicians.org
See Danny's as an example:
https://ethereum-magicians.org/t/eip-7045-increase-max-attestation-inclusion-slot/14342

status: Draft
type: Standards Track
category: Core
created: 2023-05-18
---

## Abstract

> Lock validator voluntary exit signature domain on Capella for perpetual validity. Currently, signed voluntary exits are only valid for two forks.

Reviewer 3: I think "upgrades" or "hardforks" is more specific and valuable here. as "forks" can mean manythings including short term fork-choice disagreements

```suggestion
Lock validator voluntary exit signature domain on Capella for perpetual validity. Currently, signed voluntary exits are only valid for two upgrades.
```

## Motivation

> Currently, signed voluntary exits are valid for only two forks, due to the Beacon Chain state holding only the current and previous fork. This limitation affects increases the complexity of some staking operations. Specifically those that require the exchange of signed exits ahead of time valid for an unbounded number of forks.

Reviewer: ```suggestion
Currently, signed voluntary exits are valid up-to only two forks for block inclusion due to the Beacon Chain state considering only the current and previous fork. This limitation increases the complexity of some staking operations, specifically those that require the exchange of signed exits ahead of time for an unbounded number of forks.
```

> Currently, signed voluntary exits are valid up-to only two forks for block inclusion due to the Beacon Chain state considering only the current and previous fork. This limitation increases the complexity of some staking operations, specifically those that require the exchange of signed exits ahead of time for an unbounded number of forks.

Reviewer 3: ditto

```suggestion
Currently, signed voluntary exits are valid up-to only two upgrades for block inclusion due to the Beacon Chain state considering only the current and previous fork version. This limitation increases the complexity of some staking operations, specifically those in which the staking operator (holder of active key) is distinct from the owner of the funds (holder of the withdrawal credential). Because voluntary exits can only be signed by the active key, such a relationship requires the exchange of signed exits ahead of time for an unbounded number of forks.
```

> The limited validity of volutary exits was originally motivated to isolate them in the event of a perpetual fork. If fork A and B exist and a validator operators on both, if they send an exit, it will be replayable on both. However, this positive is not sufficient to justify the UX degradation exposed above.

Reviewer: ```suggestion
The limited validity of voluntary exits was originally motivated to isolate them in the event of a perpetual fork. If fork A and B exist and a validator operates on both, if they send an exit, it will be replayable on both. However, this possibility is not sufficient to justify the UX degradation exposed above.
```

> The limited validity of voluntary exits was originally motivated to isolate them in the event of a perpetual fork. If fork A and B exist and a validator operates on both, if they send an exit, it will be replayable on both. However, this possibility is not sufficient to justify the UX degradation exposed above.

Reviewer 3: ```suggestion
The limited validity of voluntary exits was originally motivated to isolate them in the event of a hard fork that results in two maintained chains. If fork A and B exist and a validator operates on both, if they send an exit, it will be replayable on both. However, this possibility is not sufficient to justify the UX degradation exposed above, as no funds are at risk and the staker can re-stake on one or both of the chains.
```

## Specification

### Consensus Layer

> The mechanics of the specification can be found in the [Consensus Layer specs](https://github.com/ethereum/consensus-specs/pull/3288).

Reviewer: you will need to refer to a `blob` commit of the relevant content rather than a PR

Reviewer: @dapplion we are modifying the regex to support the `compare` diffs in drafts instead of PRs as the diffs are static: so basically this for your PR: `https://github.com/ethereum/consensus-specs/compare/53a9324cf0afd90ea06c5ebb0bfef8fc95a01fd1...1c35eb1c3303fe1e0b101323106d766d8f848cd6` which would is [this link](https://github.com/ethereum/consensus-specs/compare/53a9324cf0afd90ea06c5ebb0bfef8fc95a01fd1...1c35eb1c3303fe1e0b101323106d766d8f848cd6) which seems that the entire PR diff would need to be squashed in a commit for diff to be genrated correctly

> The mechanics of the specification can be found in the [Consensus Layer specs](https://github.com/ethereum/consensus-specs/commit/14212958d3605c5dd4f8ab617f157328f53ce559).

Reviewer 3: Specification changes are built into the Consensus Specs Deneb upgrade.

The specific makes one change to the state transition function:
* Modify `process_voluntary_exit`](https://github.com/ethereum/consensus-specs/blob/75971a8c218b1d76d605dd8b88a08d39c42de221/specs/deneb/beacon-chain.md#modified-process_voluntary_exit) to compute the signing domain and root fixed on `CAPELLA_FORK_VERSION`.

Additionally, the `voluntary_exit` gossip conditions are implicitly modified to support this change.

> To make the change backwards compatible the signature domain is locked on the capella fork

Reviewer 2: ```suggestion
To make the change backwards compatible the signature domain is locked on the Capella fork
```

### Execution Layer

This specification does not require any changes to the Execution Layer.

## Rationale

> Perpetually valid signed exits allow simpler staking operation designs. It also aligns the UX of such objects to BLSToExecutionChanges and deposits, such that downstream tooling do not need to be updated with fork future information.

Reviewer 3: ```suggestion
Perpetually valid signed voluntary exits allow simpler staking operation designs. It also aligns the UX of such objects to `BLSToExecutionChanges` and deposits, such that downstream tooling does not need to be updated with fork version information.
```

## Backwards Compatibility

This change is backwards compatible to the Consensus Layer of Ethereum block processing logic.

The expectation of future validity of exits is not forward compatible. Specifically, users who have already pre-signed exits utilizing the Deneb fork domain with an expectation of their validity should be aware that these pre-signed exits will no longer be recognized as valid. Consequently, users should adjust their approach moving forward. For continued validity across forks, including Deneb and subsequent forks, users should ensure that their exits are signed using the Capella fork domain.

> There are no forwards/backwards compatibility issues with the Execution Layer

Reviewer 3: ```suggestion
There are no forwards/backwards compatibility issues with the Execution Layer.
```

## Test Cases

Test cases are work-in-progress within the standard Consensus Layer tests.

## Security Considerations

> The divergent signature domains across forked networks would previously have prevented the replay of VoluntaryExits after two hard forks. This specification change causes this replay protection to no longer exist. These potential replays could impact individual stakers on both sides of a fork, but does not impact the security of the chain.

Reviewer: ```suggestion
The divergent signature domains across forked networks would previously have prevented the replay of VoluntaryExits after two hard forks. This specification change causes the replay protection to no longer exist. These potential replays could impact individual stakers on both sides of a fork, but does not impact the security of the chain.
```

> The divergent signature domains across forked networks would previously have prevented the replay of VoluntaryExits after two hard forks. This specification change causes the replay protection to no longer exist. These potential replays could impact individual stakers on both sides of a fork, but does not impact the security of the chain.

Reviewer 3: ```suggestion
The divergent signature domains across forked networks would previously have prevented the replay of VoluntaryExits after two hard forks. This specification change causes the replay protection to no longer exist. These potential replays could impact individual stakers on both sides of a fork, but does not put funds at risk and does not impact the security of the chain.
```
